File: Samples/FSharp/Streams/Chat.Client/ChatClient.fs
module Client.ChatClient

open FSharp.Control.Tasks
open Orleankka
open Orleankka.FSharp
open Messages

type ChatClient = {
   UserName: string
   User: ActorRef<ChatUserMessage>
   RoomName: string   
   Room: StreamRef<ChatRoomMessage>   
   Subscription: Option<StreamSubscription>
}

let join (client:ChatClient) = task {   
   let! sb = client.Room.Subscribe (fun message ->
      if message.UserName <> client.UserName then printfn "%s" message.Text
   )

   do! client.User <! Join(client.RoomName)
   return { client with Subscription = Some sb }
}

let leave (client:ChatClient) = task {
   do! client.Subscription.Value.Unsubscribe()
   do! client.User <! Leave(client.RoomName)
}

let say (client:ChatClient) (message:string) = task {
   do! client.User <! Say(client.RoomName, message)   
}

File: Samples/CSharp/Streams/Chat.Client/ChatClient.cs
using System;
using System.Threading.Tasks;

using Orleankka;

namespace Example
{
    class ChatClient
    {
        readonly ActorRef<IChatUser> user;
        readonly IActorSystem system;
        StreamSubscription<ChatRoomMessage> subscription;

        public ChatClient(IActorSystem system, string user, string room)
        {
            this.user = system.TypedActorOf<IChatUser>(user);
            this.system = system;
            RoomName = room;
        }

        public async Task Join()
        {
            await Subscribe();
            await user.Tell(new Join {Room = RoomName});
        }

        async Task Subscribe()
        {
            var room = system.StreamOf<ChatRoomMessage>("sms", RoomName);
            subscription = await room.Subscribe((message, _) =>
            {
                if (message.User != UserName)
                    Console.WriteLine(message.Text);
            });
        }

        public async Task Resubscribe()
        {
            while (true)
            {
                try
                {
                    await Subscribe();
                    break;
                }
                catch
                {
                    Console.WriteLine("Error resubscribing to stream. Will retry in 2 seconds");
                    await Task.Delay(2000);
                }
            }
        }

        public async Task Leave()
        {
            await subscription.Unsubscribe();
            await user.Tell(new Leave {Room = RoomName});
        }

        public async Task Say(string message)
        {
            await user.Tell(new Say {Room = RoomName, Message = message});
        }

        string UserName => user.Path.Id;
        string RoomName { get; }
    }
}
